aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,647 @@
+
+      S_1._values() = [exp(11 - maxX_1[0]) / (exp(11 - maxX_1[0]) + exp(14 - maxX_1[0]) + exp(15 - maxX_1[0])),
+                       exp(14 - maxX_1[0]) / (exp(11 - maxX_1[0]) + exp(14 - maxX_1[0]) + exp(15 - maxX_1[0])),
+                       exp(15 - maxX_1[0]) / (exp(11 - maxX_1[0]) + exp(14 - maxX_1[0]) + exp(15 - maxX_1[0])),
+                       exp(22 - maxX_1[1]) / (exp(22 - maxX_1[1]) + exp(24 - maxX_1[1])),
+                       exp(24 - maxX_1[1]) / (exp(22 - maxX_1[1]) + exp(24 - maxX_1[1]))]
+                    = [exp(-4) / (exp(-4) + exp(-1) + 1),
+                       exp(-1) / (exp(-4) + exp(-1) + 1),
+                       1 / (exp(-4) + exp(-1) + 1),
+                       exp(-2) / (exp(-2) + 1),
+                       1 / (exp(-2) + 1)]
+
+    To obtain the above via the for-loop over
+    `nnz(=len(X._values()))`, we introduce the indices mapping `pool`
+    as follows:
+
+      indices = X._indices()
+      for i in range(nnz):
+          for j in range(nnz):
+              if indices[d, i] == indices[d, j]:
+                  assert pool_d[i] == pool_d[j]
+              else:
+                  assert pool_d[i] != pool_d[j]
+
+    that is, the entries with values indices i and j are in the same
+    pool iff their locations in the grid of tensor indices align with
+    the direction along which the softmax is calculated. The `pool`
+    mapping maps the X._values() indices to the corresponding pool
+    index.
+
+    To save memory and processor resources, we pre-compute the entries
+    of maxX tensor and the sums of exponents as follows:
+
+      mx_d = [max(values[i] for i in range(nnz) if pool_0[i] == k) for k in pool_d]
+      exp_sum_d = [sum(exp(values[i] - mx_d[k]) for i in range(nnz) if pool_d[i] == k) for k in pool_d]
+
+    For example, if
+
+      pool_0 = [0, 1, 2, 3, 1]
+      pool_1 = [0, 0, 0, 1, 1]
+
+    then
+
+      mx_0 = [11, 24, 15, 22]
+      mx_1 = [15, 24]
+      exp_sum_0 = [1, (exp(-10) + 1), 1, 1]
+      exp_sum_1 = [(exp(-4) + exp(-1) + 1), (exp(-2) + 1)]
+
+    and
+
+      S_0._values() = [exp(11 - mx_0[pool_0[0]]) / exp_sum_0[pool_0[0]]
+                       exp(14 - mx_0[pool_0[1]]) / exp_sum_0[pool_0[1]]
+                       exp(15 - mx_0[pool_0[2]]) / exp_sum_0[pool_0[2]]
+                       exp(22 - mx_0[pool_0[3]]) / exp_sum_0[pool_0[3]]
+                       exp(24 - mx_0[pool_0[4]]) / exp_sum_0[pool_0[4]]
+
+    or in general,
+
+      S_d._values() = [exp(values[i] - mx_d[pool_d[i]]) / exp_sum_d[pool_d[i] for i in range(nnz)]
+
+    The above algorithm can be easily extended for cases with
+    non-scalar dense part of the sparse tensor where all scalar
+    operations become element-wise tensor operations.
+
+    The implementation below has more optimizations such as that
+    collect pool indices for enabling concurrency, minimize the calls
+    to exp functions as well as reuse of softmax implementation for
+    log_softmax.
```

Comment:
Thank you, this is a very nice comment. I wish every kernel had comments like this :)